bug: fail to push commits in GitHub Actions during `Run sync_scaffold script` step

When multiple GitHub Actions workflows are queued, the `git push` operation in the second workflow fails. This issue occurs because the second workflow does not fetch the most recent Git history after the first workflow has completed, resulting in a failure to update remote references properly.

Steps to reproduce:

1. Manually trigger the `https://github.com/Oreoxmt/pingcap-docsite-preview/actions/workflows/sync_scaffold.yml` workflow twice or more.
2. Observe the status of the second workflow after the first workflow succeeds.

Observed behavior:

The second workflow (such as [16244237197](https://github.com/Oreoxmt/pingcap-docsite-preview/actions/runs/5988751296/job/16244237197)) fails with the following error message:

```
To https://github.com/Oreoxmt/pingcap-docsite-preview
 ! [rejected]        dev -> dev (fetch first)
error: failed to push some refs to 'https://github.com/Oreoxmt/pingcap-docsite-preview'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
Error: Process completed with exit code 1.
```

--- a/.github/git_push.sh
+++ b/.github/git_push.sh
@@ -1,0 +1,47 @@
+#!/bin/bash
+
+set -e
+
+BRANCH="$1"
+
+# Try to push changes to the remote branch
+# Call handle_push_failure() if push fails
+try_push() {
+  git push || handle_push_failure
+}
+
+# Handle merge conflicts during push
+# Force push if all conflicts are in markdown-pages
+# Otherwise print error and exit
+handle_conflict() {
+  declare -a CONFLICT_FILES
+
+  while read -r file; do
+    if [[ ! "$file" =~ ^markdown-pages/ ]]; then
+      CONFLICT_FILES+=("$file")
+    fi
+  done < <(git diff --name-only "$BRANCH" "origin/$BRANCH")
+
+  if [[ ${#CONFLICT_FILES[@]} -eq 0 ]]; then
+    echo "All conflicts are in the markdown-pages folder. Force pushing."
+    git push -f
+  else
+    echo "Please resolve conflicts manually. Conflict files are:"
+    printf '%s\n' "${CONFLICT_FILES[@]}"
+    exit 1
+  fi
+}
+
+# Handle failure during git push
+# Try pull first, then push
+# If pull fails, call handle_conflict()
+handle_push_failure() {
+  if git pull origin "$BRANCH" --no-rebase; then
+    git push
+  else
+    git fetch
+    handle_conflict
+  fi
+}
+
+try_push

--- a/.github/workflows/sync_scaffold.yml
+++ b/.github/workflows/sync_scaffold.yml
@@ -5,6 +5,10 @@ on:
   schedule:
     - cron: "0 0 */15 * *" # every 15 days
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
 jobs:
 
   sync_scaffold:
@@ -21,6 +25,6 @@ jobs:
         git config user.name "Docsite Preview Bot"
         git config user.email ""
         ./sync_scaffold.sh
-        git push
+        ./.github/git_push.sh ${{ github.ref_name }}
       env:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

